src/main/java/org/craftercms/engine/util/servlet/ConfigAwareProxyServlet.java

```diff
@@ -84,4 +111,155 @@ protected void copyResponseHeaders(HttpResponse proxyResponse, HttpServletReques
         }
     }
 
+    @Override
+    protected void service(HttpServletRequest servletRequest, HttpServletResponse servletResponse)
+            throws IOException, ServletException {
+        var websocketKey = servletRequest.getHeader(WEBSOCKET_IDENTIFIER_HEADER);
+        if (websocketKey != null) {
+            websocketHandlerService(servletRequest, servletResponse);
+        } else {
+            super.service(servletRequest, servletResponse);
+        }
+    }
+
+    /**
+     * Handle the initial handshake request for websocket connection and establish the websocket.
+     * A WebSocket interaction begins with an HTTP request that uses the HTTP Upgrade header to upgrade to the WebSocket protocol.
+     * This method first forward the handshake request to the websocket server and read the response.
+     * If the response is 101, use {@link WsUpgradeHandler} to establish the websocket proxy among client <==> engine <==> websocket server.
+     * @param servletRequest servlet request from client
+     * @param servletResponse servlet response to client
+     */
+    private void websocketHandlerService(HttpServletRequest servletRequest, HttpServletResponse servletResponse) throws IOException, ServletException {
+        //initialize request attributes from caches if unset by a subclass by this point
+        if (servletRequest.getAttribute(ATTR_TARGET_URI) == null) {
+            servletRequest.setAttribute(ATTR_TARGET_URI, targetUri);
+        }
+        if (servletRequest.getAttribute(ATTR_TARGET_HOST) == null) {
+            servletRequest.setAttribute(ATTR_TARGET_HOST, targetHost);
+        }
+        String proxyRequestUri = rewriteUrlFromRequest(servletRequest);
+        URL url = new URL(proxyRequestUri);
+
+        // Create a new socket connection to the websocket server
+        Socket socketProxyClient = new Socket(url.getHost(), url.getPort());
+        boolean closeSocket = false;
+        try {
+            var socketIn = socketProxyClient.getInputStream();
+            var socketOut = socketProxyClient.getOutputStream();
+
+            // Send the handshake request from engine to the websocket server
+            String handshakeRequest = getProxyHandshakeRequest(servletRequest, url);
+            socketOut.write(handshakeRequest.getBytes(StandardCharsets.UTF_8));
+            socketOut.flush();
+
+            // Get handshake response from websocket server to engine
+            String[] responseLines = getProxyHandshakeResponse(socketIn);
```

Comment:
This seems to be unbounded. Can this result in overallocation of memory?

Reply:
The method name was incorrect so I updated it to `getProxyHandshakeResponseHeaders`. This method will only read the header of the handshake request (There is only 1 handshake request for one websocket client). In the method implementation, notice that we build a string up to 512 chars with:
```
StringBuilder responseBytes = new StringBuilder(512);
```
And break into headers Array `String[]`. HTTP headers are separated from the body with a newline character(https://developer.mozilla.org/en-US/docs/Web/HTTP/Messages#http_responses) so the implementation also guarantees that we only read the headers.

Reply:
HTTPd/Tomcat/etc will limit the HTTP Headers size.